# Lecture 3: Programming Example - Pandas Fundamentals with Washington D.C. Data

## Introduction: Your First Day as a Data Consultant

Welcome to your first hands-on session as a junior data consultant! Today, you'll work with real Washington D.C. bike-sharing data, learning pandas step-by-step. We'll start with absolute basics and build up to loading and exploring your client's dataset.

> **🚀 Interactive Learning Alert**
>
> This is a hands-on programming tutorial with code examples and challenges. For the best learning experience:
>
> - **Click "Open in Colab"** at the bottom of this notebook to run it in Google Colab
> - **Execute each code cell** by pressing **Shift + Enter** to see the results
> - **Complete the challenges** to practice what you learn
>

---

## Step 1: Setting Up Your Data Analysis Environment

Let's start by importing the tools you'll need for data analysis. Think of this like setting up your workbench before starting a project:

In [47]:
# Import pandas - your primary data manipulation tool
import pandas as pd

# Print confirmation
print("Pandas imported successfully!")
print(f"Pandas version: {pd.__version__}")

Pandas imported successfully!
Pandas version: 2.2.2


> **Note:** To run a code cell in Jupyter Notebook, click inside the cell and press **Shift + Enter**. This will execute the code and show the output directly below the cell.

**What this does:**
- `import pandas as pd` makes the pandas library available with the shorthand "pd"
- The shorthand `pd` is a universal convention - all pandas users worldwide use this
- Checking the version ensures you're working with up-to-date tools. In real-world projects, you usually just import pandas and skip printing the version or confirmation messages. We're doing it here only for teaching purposes, so you can see exactly what's happening step by step.

---

### Challenge 1: Import Practice
Import pandas yourself, then check that it worked by running `type(pd)`. The output should be `<class 'module'>`, confirming that `pd` is a module.

In [48]:
# Your code here
import pandas as pd  # Fill in the library name
print(type(pd))

<class 'module'>


> **Note:** To solve the exercise, replace each `_____` with the correct value or code. After that, expand the Solution section to compare your answer with the correct one.

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Your code here
import pandas as pd  # Fill in the library name
print(type(pd))
```

</details>

---

## Step 2: Understanding Series - Single Column Data

Let's start with Series by creating bike rental data for a typical Monday morning. Think of Series as a single column from a spreadsheet:

In [49]:
# Create a Series with hourly bike rentals
morning_rentals = pd.Series([15, 23, 45, 67, 89, 156, 234, 287])
print("Morning Bike Rentals:")
print(morning_rentals)

Morning Bike Rentals:
0     15
1     23
2     45
3     67
4     89
5    156
6    234
7    287
dtype: int64


**What this does:**
- `morning_rentals = pd.Series([...])` creates a pandas Series object and saves it in the variable `morning_rentals`
- The numbers on the right column represent bike rentals for each hour
- The numbers on the left column are index numbers that Pandas automatically assigns (0, 1, 2, etc.)

### Challenge 2: Create Your Own Series
Create a Series representing temperature readings for the same morning hours. Use these temperatures: [6, 7, 8, 10, 12, 13, 14, 16]

In [50]:
# Import pandas for Series creation
import pandas as pd

# Your code here - create a Series called 'morning_temps'
morning_temps = pd.Series([6, 7, 8, 10, 12, 13, 14, 16])  # Fill in the temperature list
print(morning_temps)  # Print your Series to verify it worked

0     6
1     7
2     8
3    10
4    12
5    13
6    14
7    16
dtype: int64


<details>
<summary>💡 <strong>Tip</strong> (click to expand)</summary>


When creating Series objects, consider these best practices:
- Use descriptive variable names like `morning_temps` instead of generic names like `data`
- Print the Series to verify it contains the expected values and structure

</details>

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Import pandas for Series creation
import pandas as pd

# Your code here - create a Series called 'morning_temps'
morning_temps = pd.Series([6, 7, 8, 10, 12, 13, 14, 16])  # Fill in the temperature list
print(morning_temps)  # Print your Series to verify it worked
```

</details>

---

## Step 3: Adding Meaningful Labels to Series

Raw index numbers (0, 1, 2) aren't very business-friendly. Let's add meaningful labels:

In [51]:
# Create Series with meaningful hour labels
hour_labels = ['6 AM', '7 AM', '8 AM', '9 AM', '10 AM', '11 AM', '12 PM', '1 PM']
morning_rentals_labeled = pd.Series([15, 23, 45, 67, 89, 156, 234, 287],
                                   index=hour_labels)
print("Labeled Morning Bike Rentals:")
print(morning_rentals_labeled)

Labeled Morning Bike Rentals:
6 AM      15
7 AM      23
8 AM      45
9 AM      67
10 AM     89
11 AM    156
12 PM    234
1 PM     287
dtype: int64


**What this does:**
- `hour_labels = ['6 AM', '7 AM', '8 AM', '9 AM', '10 AM', '11 AM', '12 PM', '1 PM']` creates a list of time labels that represent each hour in the morning
- `index=hour_labels` replaces default numbers with business-meaningful labels created in the previous line
- Now each rental count is clearly connected to its time period

Now when you show this to your client, they immediately understand that 1 PM has the highest rentals (287), which makes perfect sense for lunch-time bike usage.

### Challenge 3: Access Specific Data Points
Using your labeled Series, find the bike rentals at 9 AM. Use this syntax: `morning_rentals_labeled['9 AM']`

In [52]:
# Import pandas and create the labeled Series
import pandas as pd

# Create Series with meaningful hour labels
hour_labels = ['6 AM', '7 AM', '8 AM', '9 AM', '10 AM', '11 AM', '12 PM', '1 PM']
morning_rentals_labeled = pd.Series([15, 23, 45, 67, 89, 156, 234, 287],
                                   index=hour_labels)

# Your code here - access the 9 AM value from the labeled Series
nine_am_rentals = morning_rentals_labeled['9 AM']  # Fill in the correct label
print(f"Bike rentals at 9 AM: {nine_am_rentals }")  # Print the result

Bike rentals at 9 AM: 67


<details>
<summary>💡 <strong>Tip</strong> (click to expand)</summary>


When accessing Series data by label, keep these techniques in mind:
- In this challenge, we're using square brackets with the exact label: `series_name['label']`
- Check available labels with `series_name.index` to see all options
- Access multiple values: `series_name[['9 AM', '10 AM']]` (note double brackets)
- Use `.loc[]` for explicit label-based selection: `series_name.loc['9 AM']`
- Be careful with exact spelling and spacing in labels to avoid KeyError

</details>

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Import pandas and create the labeled Series
import pandas as pd

# Create Series with meaningful hour labels
hour_labels = ['6 AM', '7 AM', '8 AM', '9 AM', '10 AM', '11 AM', '12 PM', '1 PM']
morning_rentals_labeled = pd.Series([15, 23, 45, 67, 89, 156, 234, 287],
                                   index=hour_labels)

# Your code here - access the 9 AM value from the labeled Series
nine_am_rentals = morning_rentals_labeled['9 AM']  # Fill in the correct label
print(f"Bike rentals at 9 AM: {nine_am_rentals}")  # Print the result
```

</details>

---

## Step 4: Creating Your First DataFrame - Complete Business Data

Now let's combine multiple pieces of information into a DataFrame. Think of this as creating a complete spreadsheet with multiple columns:

In [53]:
# Create a comprehensive DataFrame with multiple variables
bike_operations_data = pd.DataFrame({
    'hour': ['6 AM', '7 AM', '8 AM', '9 AM', '10 AM'],
    'temperature': [5.6, 6.7, 8.3, 10.0, 11.7],
    'bike_rentals': [15, 23, 45, 67, 89],
    'weather_condition': ['Clear', 'Clear', 'Partly Cloudy', 'Clear', 'Clear']
})

print("Complete Bike Operations Data:")
print(bike_operations_data)

Complete Bike Operations Data:
    hour  temperature  bike_rentals weather_condition
0   6 AM          5.6            15             Clear
1   7 AM          6.7            23             Clear
2   8 AM          8.3            45     Partly Cloudy
3   9 AM         10.0            67             Clear
4  10 AM         11.7            89             Clear


**What this does:**
- `pd.DataFrame({...})` creates a DataFrame with multiple columns
- The DataFrame constructor accepts a dictionary where:
  - each key (like 'hour', 'temperature', etc.) becomes a column name
  - each list in the dictionary provides the values for that column
- All rows stay aligned (first hour corresponds to first temperature, etc.)

You can immediately see patterns - bike rentals increase with temperature and time, giving your client valuable operational insights..

### Challenge 4: Add a New Column
Add a column called 'user_satisfaction' with values [3.2, 3.5, 3.8, 4.1, 4.3] representing customer satisfaction ratings.

In [54]:
# Import pandas and create the DataFrame
import pandas as pd

# Create a comprehensive DataFrame with multiple variables
bike_operations_data = pd.DataFrame({
    'hour': ['6 AM', '7 AM', '8 AM', '9 AM', '10 AM'],
    'temperature': [5.6, 6.7, 8.3, 10.0, 11.7],
    'bike_rentals': [15, 23, 45, 67, 89],
    'weather_condition': ['Clear', 'Clear', 'Partly Cloudy', 'Clear', 'Clear']
})

# Your code here - add a new column with satisfaction ratings
bike_operations_data['user_satisfaction'] = [3.2, 3.5, 3.8, 4.1, 4.3]  # Fill in column name
print(bike_operations_data)  # Print the updated DataFrame
print(bike_operations_data.dtypes)

    hour  temperature  bike_rentals weather_condition  user_satisfaction
0   6 AM          5.6            15             Clear                3.2
1   7 AM          6.7            23             Clear                3.5
2   8 AM          8.3            45     Partly Cloudy                3.8
3   9 AM         10.0            67             Clear                4.1
4  10 AM         11.7            89             Clear                4.3
hour                  object
temperature          float64
bike_rentals           int64
weather_condition     object
user_satisfaction    float64
dtype: object


<details>
<summary>💡 <strong>Tip</strong> (click to expand)</summary>


When adding new columns to DataFrames, follow these best practices:
- Ensure the new data list has the same length as existing rows: `len(new_data) == len(df)`
- Use descriptive column names that clearly indicate what the data represents
- Verify the addition worked: `df.columns` shows all column names including the new one
- Check data types: `df.dtypes` to ensure the new column has appropriate type (float64 for ratings)
- You can also add columns using `df.assign(column_name=values)` for a more functional approach

</details>

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Import pandas and create the DataFrame
import pandas as pd

# Create a comprehensive DataFrame with multiple variables
bike_operations_data = pd.DataFrame({
    'hour': ['6 AM', '7 AM', '8 AM', '9 AM', '10 AM'],
    'temperature': [5.6, 6.7, 8.3, 10.0, 11.7],
    'bike_rentals': [15, 23, 45, 67, 89],
    'weather_condition': ['Clear', 'Clear', 'Partly Cloudy', 'Clear', 'Clear']
})

# Your code here - add a new column with satisfaction ratings
bike_operations_data['user_satisfaction'] = [3.2, 3.5, 3.8, 4.1, 4.3]  # Fill in column name
print(bike_operations_data)  # Print the updated DataFrame
```

</details>

---

## Step 5: Loading Real Washington D.C. Dataset

Now for the real challenge - loading your client's actual historical data. This is where professional consulting begins:

In [55]:
# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Confirm successful loading
print(f"Dataset loaded successfully!")
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

Dataset loaded successfully!
Shape: 10886 rows × 12 columns


**What this does:**
- `pd.read_csv()` reads data from a CSV file into a DataFrame
- CSV (Comma-Separated Values) a standard format for sharing tabular data
- The shape tells you how much data you have to work with

Always check the shape immediately after loading - it confirms the file loaded correctly and gives you a sense of your dataset size.

### Challenge 5: Explore the Column Names
Print the column names using `df.columns` to see what variables are available in your dataset.

In [56]:
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - explore the available columns
print("Available columns:")
print(list(df.columns))  # Fill in the DataFrame name

Available columns:
['datetime', 'season', 'holiday', 'workingday', 'weather', 'temp', 'atemp', 'humidity', 'windspeed', 'casual', 'registered', 'count']


<details>
<summary>💡 <strong>Tip</strong> (click to expand)</summary>


When exploring column names in a new dataset, use these investigation techniques:
- `df.columns` returns an Index object with all column names
- `list(df.columns)` converts to a regular Python list for easier reading
- `len(df.columns)` tells you how many variables you have to work with
- Look for patterns in naming conventions to understand data structure

</details>

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - explore the available columns
print("Available columns:")
print(list(df.columns))  # Fill in the DataFrame name
```

</details>

---

## Step 6: First Look at Real Transportation Data

Let's examine the first few rows to understand the data structure:

In [57]:
# Display the first 5 rows
print("First 5 rows of the dataset:")
print(df.head())

First 5 rows of the dataset:
              datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   
2  2011-01-01 02:00:00       1        0           0        1  9.02  13.635   
3  2011-01-01 03:00:00       1        0           0        1  9.84  14.395   
4  2011-01-01 04:00:00       1        0           0        1  9.84  14.395   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40  
2        80        0.0       5          27     32  
3        75        0.0       3          10     13  
4        75        0.0       0           1      1  


**What this shows:**
- `head()` displays the first 5 rows by default (you can specify a different number like `df.head(10)` to see 10 rows)
- You'll see actual bike-sharing data with timestamps, weather, and usage counts
- Each row represents one hour of bike-sharing operations

**Understanding the real data:**
- `datetime`: When this data was recorded
- `season`, `holiday`, `workingday`: Operational context
- `weather`, `temp`, `humidity`, `windspeed`: Weather conditions
- `casual`, `registered`, `count`: Different types of users and total rentals

This is the foundation of all your future analysis for this client.

### Challenge 6: Look at the Last Few Rows
Use `df.tail()` to see the last 5 rows of the dataset. This helps verify you have complete data coverage.

In [58]:
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - examine the last few rows
print("Last 5 rows of the dataset:")
print(df.tail)  # Fill in the DataFrame name and method

Last 5 rows of the dataset:
<bound method NDFrame.tail of                   datetime  season  holiday  workingday  weather   temp  \
0      2011-01-01 00:00:00       1        0           0        1   9.84   
1      2011-01-01 01:00:00       1        0           0        1   9.02   
2      2011-01-01 02:00:00       1        0           0        1   9.02   
3      2011-01-01 03:00:00       1        0           0        1   9.84   
4      2011-01-01 04:00:00       1        0           0        1   9.84   
...                    ...     ...      ...         ...      ...    ...   
10881  2012-12-19 19:00:00       4        0           1        1  15.58   
10882  2012-12-19 20:00:00       4        0           1        1  14.76   
10883  2012-12-19 21:00:00       4        0           1        1  13.94   
10884  2012-12-19 22:00:00       4        0           1        1  13.94   
10885  2012-12-19 23:00:00       4        0           1        1  13.12   

        atemp  humidity  windspeed  casua

<details>
<summary>💡 <strong>Tip</strong> (click to expand)</summary>


When examining the end of your dataset, consider these data quality checks:
- Compare last row's datetime to first row's datetime to understand time coverage
- Check if the last rows have complete data or if there are missing values
- Use `df.tail(10)` to see more rows if you want a larger sample
- Look for any unusual patterns or data entry errors in the final records
- Verify that the data collection didn't stop abruptly mid-period

</details>

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - examine the last few rows
print("Last 5 rows of the dataset:")
print(df.tail())  # Fill in the DataFrame name and method
```

</details>

---

## Step 7: Understanding Your Dataset Size and Structure

Professional data analysis requires understanding exactly what you're working with:

In [59]:
# Get detailed information about the dataset
print("Dataset Information:")
print(f"Total records: {len(df)}")
print(f"Total variables: {len(df.columns)}")

# Show data types for each column
print("\nData Types:")
print(df.dtypes)

Dataset Information:
Total records: 10886
Total variables: 12

Data Types:
datetime       object
season          int64
holiday         int64
workingday      int64
weather         int64
temp          float64
atemp         float64
humidity        int64
windspeed     float64
casual          int64
registered      int64
count           int64
dtype: object


**What this tells you:**
- **Total records**: How much historical data your client has collected
- **Total variables**: How many different factors you can analyze
- **Data types**: What kind of analysis you can perform on each variable

More historical data means more reliable predictions. The variety of variables (weather, time, user types) means you can build sophisticated demand forecasting models.

### Challenge 7: Calculate Time Coverage
The dataset contains hourly data. Calculate how many days of data you have by dividing the total rows by 24.

In [60]:
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - calculate days of coverage
total_days = len(df) / 24 # Fill in DataFrame name and divisor
print(f"Dataset covers approximately {total_days:.1f} days")  # Fill in variable name

Dataset covers approximately 453.6 days


<details>
<summary>💡 <strong>Tip</strong> (click to expand)</summary>


When calculating time coverage for time series data, consider these analysis approaches:
- Use `:.1f` formatting to display days with one decimal place for readability
- Calculate weeks as well: `total_days / 7` for business planning context
- Consider if you have complete days: `len(df) % 24` shows any partial days

</details>

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - calculate days of coverage
total_days = len(df) / 24  # Fill in DataFrame name and divisor
print(f"Dataset covers approximately {total_days:.1f} days")  # Fill in variable name
```

</details>

---

## Step 8: Basic Statistical Summary

Understanding the data distribution helps identify patterns and potential issues:

In [61]:
# Generate statistical summary for numerical variables
print("Statistical Summary:")
print(df.describe())

Statistical Summary:
             season       holiday    workingday       weather         temp  \
count  10886.000000  10886.000000  10886.000000  10886.000000  10886.00000   
mean       2.506614      0.028569      0.680875      1.418427     20.23086   
std        1.116174      0.166599      0.466159      0.633839      7.79159   
min        1.000000      0.000000      0.000000      1.000000      0.82000   
25%        2.000000      0.000000      0.000000      1.000000     13.94000   
50%        3.000000      0.000000      1.000000      1.000000     20.50000   
75%        4.000000      0.000000      1.000000      2.000000     26.24000   
max        4.000000      1.000000      1.000000      4.000000     41.00000   

              atemp      humidity     windspeed        casual    registered  \
count  10886.000000  10886.000000  10886.000000  10886.000000  10886.000000   
mean      23.655084     61.886460     12.799395     36.021955    155.552177   
std        8.474601     19.245033      

**What this shows:**
- **count**: How many non-missing values exist for each variable
- **mean**: Average values (useful for understanding typical conditions)
- **std**: Standard deviation (shows how much values vary)
- **min/max**: Range of values (helps identify outliers or impossible values)
- **25%, 50%, 75%**: Quartiles (show data distribution)

If minimum bike counts are 1 and maximum is 977, that's a huge range! This suggests your client experiences very different demand conditions that you'll need to understand and predict.

### Challenge 8: Focus on Key Business Metrics
Create a summary focusing only on the key business variables: temperature (`temp`), humidity (`humidity`), and bike count (`count`).

In [62]:
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - create a summary of key metrics
key_metrics = df[["temp", "humidity", "count"]].describe()  # Fill in DataFrame and column names
print("Key Business Metrics Summary:")
print(key_metrics)  # Print the summary

Key Business Metrics Summary:
              temp      humidity         count
count  10886.00000  10886.000000  10886.000000
mean      20.23086     61.886460    191.574132
std        7.79159     19.245033    181.144454
min        0.82000      0.000000      1.000000
25%       13.94000     47.000000     42.000000
50%       20.50000     62.000000    145.000000
75%       26.24000     77.000000    284.000000
max       41.00000    100.000000    977.000000


<details>
<summary>💡 <strong>Tip</strong> (click to expand)</summary>


When focusing on specific business metrics, use these analytical techniques:
- Select columns using double brackets: `df[['col1', 'col2']]` to maintain DataFrame structure
- Compare ranges across variables to understand which have more variation
- Check for outliers: are max values realistic or potentially data entry errors?
- Consider business thresholds: what temperature ranges are most relevant for operations?

</details>

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - create a summary of key metrics
key_metrics = df[['temp', 'humidity', 'count']].describe()  # Fill in DataFrame and column names
print("Key Business Metrics Summary:")
print(key_metrics)  # Print the summary
```

</details>

---

## Step 9: Selecting Specific Data for Analysis

Often you need to focus on specific parts of your dataset. Let's learn several ways to select data:

In [63]:
# Select a single column (bike counts)
bike_counts = df['count']
print(f"Bike counts - Type: {type(bike_counts)}")
print(f"Average daily rentals: {bike_counts.mean():.1f}")

# Select multiple columns for weather analysis
weather_data = df[['temp', 'humidity', 'windspeed']]
print(f"\nWeather data shape: {weather_data.shape}")
print(weather_data.head(3))

# Select first 12 rows for initial analysis
sample_data = df.head(12)
print(f"\nSample data covers first {len(sample_data)} hours")
print(sample_data)

Bike counts - Type: <class 'pandas.core.series.Series'>
Average daily rentals: 191.6

Weather data shape: (10886, 3)
   temp  humidity  windspeed
0  9.84        81        0.0
1  9.02        80        0.0
2  9.02        80        0.0

Sample data covers first 12 hours
               datetime  season  holiday  workingday  weather   temp   atemp  \
0   2011-01-01 00:00:00       1        0           0        1   9.84  14.395   
1   2011-01-01 01:00:00       1        0           0        1   9.02  13.635   
2   2011-01-01 02:00:00       1        0           0        1   9.02  13.635   
3   2011-01-01 03:00:00       1        0           0        1   9.84  14.395   
4   2011-01-01 04:00:00       1        0           0        1   9.84  14.395   
5   2011-01-01 05:00:00       1        0           0        2   9.84  12.880   
6   2011-01-01 06:00:00       1        0           0        1   9.02  13.635   
7   2011-01-01 07:00:00       1        0           0        1   8.20  12.880   
8   2011-01-

**What this demonstrates:**
- **Single column selection**: Returns a Series (one-dimensional)
- **Multiple column selection**: Returns a DataFrame (two-dimensional)
- **Row selection**: Gets a subset of the full dataset

You might analyze just weather data to understand seasonal patterns, or focus on the first few months to understand how the bike-sharing system performed during its early operations.

### Challenge 9: Create a Sample Dataset
Select only the columns 'datetime', 'temp', 'count' and only the first 168 rows (first week of data).

In [64]:
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - create a focused dataset for rush hour analysis
rush_hour_analysis = df[[ 'datetime', 'temp', 'count']].head(168)  # Fill in details
print(f"Rush hour dataset shape: {rush_hour_analysis.shape}")  # Fill in variable name
print(rush_hour_analysis)  # Print sample data

Rush hour dataset shape: (168, 3)
                datetime  temp  count
0    2011-01-01 00:00:00  9.84     16
1    2011-01-01 01:00:00  9.02     40
2    2011-01-01 02:00:00  9.02     32
3    2011-01-01 03:00:00  9.84     13
4    2011-01-01 04:00:00  9.84      1
..                   ...   ...    ...
163  2011-01-08 02:00:00  7.38     16
164  2011-01-08 03:00:00  7.38      7
165  2011-01-08 04:00:00  7.38      1
166  2011-01-08 05:00:00  6.56      5
167  2011-01-08 06:00:00  6.56      2

[168 rows x 3 columns]


<details>
<summary>💡 <strong>Tip</strong> (click to expand)</summary>


When creating focused analysis datasets, use these selection strategies:
- Verify the subset size: 168 hours = 7 days × 24 hours = 1 week
- Combine column and row selection: `df[['col1', 'col2']].head(n)`
- Confirm time coverage: compare first and last datetime values - as you will see, it doesn't match (this may mean that some hour registers are missing)

</details>

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - create a focused dataset for rush hour analysis
rush_hour_analysis = df[['datetime', 'temp', 'count']].head(168)  # Fill in details
print(f"Rush hour dataset shape: {rush_hour_analysis.shape}")  # Fill in variable name
print(rush_hour_analysis)  # Print sample data
```

</details>

---

## Step 10: Understanding Time-Based Data

Transportation data is inherently time-based. Let's work with the datetime information:

In [65]:
# Convert datetime column to pandas datetime format
df['datetime'] = pd.to_datetime(df['datetime'])
print(f"Datetime conversion successful. Type: {df['datetime'].dtype}")

# Ensure chronological order for time-based operations
df = df.sort_values('datetime').reset_index(drop=True)

# Extract useful time components
df['hour'] = df['datetime'].dt.hour
df['day_of_week'] = df['datetime'].dt.day_name()
df['month'] = df['datetime'].dt.month

# Show the first few rows with time components
print("\nData with extracted time components:")
print(df[['datetime', 'hour', 'day_of_week', 'month', 'count']].head())

Datetime conversion successful. Type: datetime64[ns]

Data with extracted time components:
             datetime  hour day_of_week  month  count
0 2011-01-01 00:00:00     0    Saturday      1     16
1 2011-01-01 01:00:00     1    Saturday      1     40
2 2011-01-01 02:00:00     2    Saturday      1     32
3 2011-01-01 03:00:00     3    Saturday      1     13
4 2011-01-01 04:00:00     4    Saturday      1      1


> Timedelta primer: A Timedelta represents a duration (difference between two timestamps). You'll use `pd.Timedelta(hours=1)` in the next step to flag gaps larger than one hour between records.

**What this accomplishes:**
- Converts text dates to actual datetime objects for analysis
- Extracts hour, day, and month for business analysis
- Enables time-based filtering and grouping

These **time-based features** unlock powerful insights. **Hour analysis** allows you to identify peak usage times for bike rebalancing, while **day analysis** helps compare weekday vs. weekend patterns. Additionally, **month analysis** reveals seasonal trends that are crucial for capacity planning.

### Challenge 10: Find Peak Hour
Use the new `hour` column to find which hour of the day has the highest average bike rentals.

In [66]:
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Convert datetime column and extract time components
df['datetime'] = pd.to_datetime(df['datetime'])
df['hour'] = df['datetime'].dt.hour
df['day_of_week'] = df['datetime'].dt.day_name()
df['month'] = df['datetime'].dt.month

# Your code here - find the peak hour for bike rentals
hourly_average = df.groupby('hour')['count'].mean()  # Fill in DataFrame, grouping column, target column
peak_hour = hourly_average.idxmax() # Fill in method to find maximum index
peak_rentals = hourly_average.max()    # Fill in method to get maximum value

print(f"Peak hour: {peak_hour}:00")  # Fill in variable name
print(f"Average rentals during peak hour: {peak_rentals:.1f}")  # Fill in variable name

Peak hour: 17:00
Average rentals during peak hour: 468.8


<details>
<summary>💡 <strong>Tip</strong> (click to expand)</summary>


When analyzing time-based patterns, use these grouping and aggregation techniques:
- `df.groupby('hour')['count'].mean()` calculates average by hour
- Use `.idxmax()` to find the index (hour) with maximum value
- Use `.max()` to get the actual maximum value
- Explore other time patterns: `df.groupby('day_of_week')['count'].mean()`
- Consider multiple aggregations: `df.groupby('hour')['count'].agg(['mean', 'std', 'count'])`
- Sort results for easier interpretation: `hourly_average.sort_values(ascending=False)`

</details>

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Convert datetime column and extract time components
df['datetime'] = pd.to_datetime(df['datetime'])
df['hour'] = df['datetime'].dt.hour
df['day_of_week'] = df['datetime'].dt.day_name()
df['month'] = df['datetime'].dt.month

# Your code here - find the peak hour for bike rentals
hourly_average = df.groupby('hour')['count'].mean()  # Fill in DataFrame, grouping column, target column
peak_hour = hourly_average.idxmax()  # Fill in method to find maximum index
peak_rentals = hourly_average.max()  # Fill in method to get maximum value

print(f"Peak hour: {peak_hour}:00")  # Fill in variable name
print(f"Average rentals during peak hour: {peak_rentals:.1f}")  # Fill in variable name
```

</details>

---

## Step 11: Data Quality Assessment - Missing Data Detection (Empty Cells)

Real-world data often has missing values (empty cells). Let's check if any values are missing from our dataset. Note that we're checking for empty cells here - detecting missing time periods (like skipped hours) requires a different approach that we'll handle in Challenge 11.

In [67]:
# Check for missing data in each column
missing_data = df.isnull().sum()
print("Missing Data Summary:")
print(missing_data)

# Calculate percentage of missing data
missing_percentage = (missing_data / len(df)) * 100
print("\nMissing Data Percentages:")
for column in df.columns:
    if missing_data[column] > 0:
        print(f"{column}: {missing_percentage[column]:.1f}%")
    else:
        print(f"{column}: 0% (Complete)")

Missing Data Summary:
datetime       0
season         0
holiday        0
workingday     0
weather        0
temp           0
atemp          0
humidity       0
windspeed      0
casual         0
registered     0
count          0
hour           0
day_of_week    0
month          0
dtype: int64

Missing Data Percentages:
datetime: 0% (Complete)
season: 0% (Complete)
holiday: 0% (Complete)
workingday: 0% (Complete)
weather: 0% (Complete)
temp: 0% (Complete)
atemp: 0% (Complete)
humidity: 0% (Complete)
windspeed: 0% (Complete)
casual: 0% (Complete)
registered: 0% (Complete)
count: 0% (Complete)
hour: 0% (Complete)
day_of_week: 0% (Complete)
month: 0% (Complete)


**What this code accomplishes:**
- `df.isnull()` creates a DataFrame of True/False values where True indicates a missing value
- `.sum()` counts the True values (which represent missing entries) for each column - since True is treated as 1 and False as 0 in arithmetic operations
- The percentage calculation shows the proportion of missing data relative to total records
- The loop displays results in a business-friendly format, clearly marking complete vs incomplete columns
- Converting to percentages helps understand the severity of missing data

By running this code, we can see that there is **no missing data** in the dataset - all columns are complete. This means that **we can confidently use any column** for calculations without data quality concerns. **If we had missing data**, we would need to decide what to do with it: drop incomplete rows, exclude unreliable columns, or fill gaps using estimates (like averages or trends) - each choice would impact the quality of our analysis.

### Challenge 11: Identify Missing Datetime Gaps (Skipped Hours)
Find where the gap between consecutive `datetime` values is greater than 1 hour — these indicate missing hourly records in the time series. This approach works because time series data should have consistent intervals. In our case, records are generated hourly, so any gap larger than 1 hour reveals missing records in the sequence.

In [68]:
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - detect gaps > 1 hour in the datetime sequence
df['datetime'] = pd.to_datetime(df['datetime'])
df = df.sort_values(by='datetime').reset_index(drop=True)  # Fill in the time column

time_diffs = df['datetime'].diff()  # Fill in the time column
gaps = df[time_diffs > pd.Timedelta(hours=1)]  # Fill in the gap size

print("Rows that follow a gap > 1 hour:")
print(gaps)  # Fill in variable name

Rows that follow a gap > 1 hour:
                 datetime  season  holiday  workingday  weather   temp  \
29    2011-01-02 06:00:00       1        0           0        3  17.22   
49    2011-01-03 04:00:00       1        0           1        1   6.56   
72    2011-01-04 04:00:00       1        0           1        1   5.74   
95    2011-01-05 04:00:00       1        0           1        1   9.84   
118   2011-01-06 04:00:00       1        0           1        2   6.56   
...                   ...     ...      ...         ...      ...    ...   
9063  2012-09-01 00:00:00       3        0           0        1  30.34   
9519  2012-10-01 00:00:00       4        0           1        1  18.86   
9975  2012-11-01 00:00:00       4        0           1        1  14.76   
10146 2012-11-08 04:00:00       4        0           1        2  12.30   
10430 2012-12-01 00:00:00       4        0           0        1  10.66   

        atemp  humidity  windspeed  casual  registered  count  
29     21.210 

<details>
<summary>💡 <strong>Tip</strong> (click to expand)</summary>


When detecting missing time periods, keep these practices in mind:
- Convert the datetime column to pandas datetime format by using `pd.to_datetime()`
- Ensure the data is sorted by time before calling `.diff()`
- Remember that `.diff()` calculates the time difference between each row and the previous row
- Use `pd.Timedelta(hours=1)` for a clear 1-hour threshold

</details>

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Your code here - detect gaps > 1 hour in the datetime sequence
df['datetime'] = pd.to_datetime(df['datetime'])
df = df.sort_values(by='datetime').reset_index(drop=True)  # Fill in the time column

time_diffs = df['datetime'].diff()  # Fill in the time column
gaps = df[time_diffs > pd.Timedelta(hours=1)]  # Fill in the gap size

print("Rows that follow a gap > 1 hour:")
print(gaps)  # Fill in variable name
```

</details>

---

## Step 12: Basic Data Filtering for Business Insights

Now we'll use data filtering to answer critical business questions that help optimize bike-sharing operations. By creating targeted subsets of our data, we can identify peak demand periods, understand weather impacts on ridership, and develop insights for operational planning:

In [69]:
# Find high-demand periods (above average usage)
average_rentals = df['count'].mean()
high_demand = df[df['count'] > average_rentals]
print(f"High-demand periods: {len(high_demand)} out of {len(df)} total hours")
print(f"That's {len(high_demand)/len(df)*100:.1f}% of all hours")

# Find cold weather operations (temperature below 8°C)
cold_weather = df[df['temp'] < 8]
print(f"\nCold weather operations: {len(cold_weather)} hours")
print(f"Average rentals in cold weather: {cold_weather['count'].mean():.1f}")

# Compare to warm weather (temperature above 26°C)
warm_weather = df[df['temp'] >= 26]
print(f"Average rentals in warm weather: {warm_weather['count'].mean():.1f}")

High-demand periods: 4356 out of 10886 total hours
That's 40.0% of all hours

Cold weather operations: 488 hours
Average rentals in cold weather: 60.8
Average rentals in warm weather: 277.2


The analysis reveals clear patterns in bike-sharing demand:

- Out of more than **10,000 recorded hours**, around **40% qualify as high-demand periods**, showing that elevated usage is a regular occurrence rather than an exception
- **Weather plays a particularly strong role**: when temperatures drop below **8°C**, bike rentals average only about **61 per hour**, reflecting how cold conditions discourage riders
- In contrast, when the temperature climbs to **26°C or higher**, average rentals surge to over **277 per hour**—more than **four times the cold-weather figure**

As you can imagine, these findings have significant implications for operational planning and resource allocation.

### Challenge 12: Weekend vs. Weekday Analysis
Filter the data to compare average bike rentals on weekends (Saturday, Sunday) versus weekdays.

In [73]:
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Convert datetime column and extract time components
df['datetime'] = pd.to_datetime(df['datetime'])
df['hour'] = df['datetime'].dt.hour
df['day_of_week'] = df['datetime'].dt.day_name()
df['month'] = df['datetime'].dt.month

# Your code here - compare weekend vs weekday bike usage
weekend_data = df[df['day_of_week'].isin(['Saturday', 'Sunday'])]  # Fill in details
weekday_data = df[~df['day_of_week'].isin(['Saturday', 'Sunday'])] # Fill in details

weekend_avg = weekend_data['count'].mean()  # Fill in column name
weekday_avg = weekday_data['count'].mean()  # Fill in column name

print(f"Weekend average rentals: {weekend_avg:.1f}")  # Fill in variable name
print(f"Weekday average rentals: {weekday_avg:.1f}")  # Fill in variable name
print(f" ABS Difference: {abs(weekend_avg - weekday_avg):.1f} rentals per hour")  # Fill in variable names
print(f" % Difference: {abs((weekend_avg - weekday_avg) / weekday_avg) * 100:.1f}% dif in rentals per hour")  # Fill in variable names

Weekend average rentals: 188.8
Weekday average rentals: 192.7
 ABS Difference: 4.0 rentals per hour
 % Difference: 2.1% dif in rentals per hour


<details>
<summary>💡 <strong>Tip</strong> (click to expand)</summary>


When comparing categorical groups like weekends vs weekdays, use these filtering strategies:
- Use `.isin(['value1', 'value2'])` to match multiple values
- Use `~` (tilde) for "not" to get the inverse: `~df['col'].isin(values)`
- Alternative approach: `df['day_of_week'].str.contains('Saturday|Sunday')`
- Consider statistical significance: do the groups have meaningful differences?
- Calculate percentage difference: `((weekend_avg - weekday_avg) / weekday_avg) * 100`
- Explore within-group variation: `weekend_data['count'].std()` vs `weekday_data['count'].std()`

</details>

<details>
<summary>🤫 <strong>Solution</strong> (click to expand)</summary>

```python
# Import required libraries and load data
import pandas as pd

# Load the Washington D.C. bike-sharing dataset
data_file_path = "https://raw.githubusercontent.com/pmarcelino/predictive-modeling/main/datasets/dataset.csv"
df = pd.read_csv(data_file_path)

# Convert datetime column and extract time components
df['datetime'] = pd.to_datetime(df['datetime'])
df['hour'] = df['datetime'].dt.hour
df['day_of_week'] = df['datetime'].dt.day_name()
df['month'] = df['datetime'].dt.month

# Your code here - compare weekend vs weekday bike usage
weekend_data = df[df['day_of_week'].isin(['Saturday', 'Sunday'])]  # Fill in details
weekday_data = df[~df['day_of_week'].isin(['Saturday', 'Sunday'])]  # Fill in details

weekend_avg = weekend_data['count'].mean()  # Fill in column name
weekday_avg = weekday_data['count'].mean()  # Fill in column name

print(f"Weekend average rentals: {weekend_avg:.1f}")  # Fill in variable name
print(f"Weekday average rentals: {weekday_avg:.1f}")  # Fill in variable name
print(f"Difference: {abs(weekend_avg - weekday_avg):.1f} rentals per hour")  # Fill in variable names
```

</details>

---

## Summary: Professional Pandas Data Analysis Fundamentals

**What We've Accomplished**:
- Established comprehensive pandas environment and data manipulation workflows
- Implemented systematic data loading and exploration methodologies for real transportation data
- Performed data quality assessment with missing data detection protocols
- Created time-based feature extraction and business intelligence filtering frameworks

**Key Technical Skills Mastered**:
- Series and DataFrame creation with meaningful business labeling systems
- CSV data loading and basic analysis for professional client datasets
- Temporal data manipulation with datetime extraction and grouping operations
- Data filtering and aggregation techniques for business insight generation

**Next Steps**: Next, we'll advance to professional data cleaning techniques, mastering missing value handling, outlier identification, and data preparation protocols that ensure our transportation datasets meet the rigorous quality standards required for sophisticated predictive modeling and client-ready analysis.

Your bike-sharing client now has a solid data foundation built with professional pandas techniques that demonstrate systematic data exploration and business-focused analytical thinking - the core competencies that consulting firms expect from junior transportation data analysts!